File: src/main/java/ar/edu/unq/spring/persistence/dto/PersonajeJPADTO.java
package ar.edu.unq.spring.persistence.dto;

import jakarta.persistence.*;
import lombok.*;

import java.util.HashSet;
import java.util.Set;

@Setter
@Getter
@ToString
@NoArgsConstructor(access = AccessLevel.PROTECTED)

@Entity
public class PersonajeJPADTO {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 500, unique = true)
    private String nombre;
    private int vida;
    private int pesoMaximo;

    @OneToMany(
            mappedBy = "owner",
            cascade = CascadeType.ALL,
            fetch = FetchType.EAGER
    )
    private Set<ItemJPADTO> inventario = new HashSet<>();


}

File: src/main/kotlin/ar/edu/unq/spring/persistence/PersonajeJPADTO.kt
package ar.edu.unq.spring.persistence

import ar.edu.unq.spring.modelo.Personaje
import javax.persistence.*

@Entity(name = "Personaje")
class PersonajeJPADTO() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
    @Column(nullable = false, length = 500, unique = true)
    var nombre: String? = null
    var vida: Int = 0
    var pesoMaximo: Int = 0

    @OneToMany(
        mappedBy = "owner",
        cascade = [CascadeType.MERGE, CascadeType.REMOVE, CascadeType.REFRESH, CascadeType.DETACH],
        fetch = FetchType.EAGER
    )
    var inventario: MutableSet<ItemJPADTO> = HashSet()

    companion object {
        fun desdeModelo(personaje: Personaje): PersonajeJPADTO{
            val dto = PersonajeJPADTO()
            dto.vida = personaje.vida
            dto.nombre = personaje.nombre
            dto.pesoMaximo = personaje.pesoMaximo
            dto.id = personaje.id
            dto.inventario = personaje.inventario.map{i -> ItemJPADTO.desdeModelo(i, dto)}.toMutableSet()
            return dto
        }
    }

    fun aModelo() : Personaje {
        val personaje =  Personaje(nombre!!)
        personaje.vida = vida
        personaje.pesoMaximo = pesoMaximo
        personaje.inventario = inventario.map{i -> i.aModelo(personaje)}.toMutableSet()
        personaje.id = id
        return personaje
    }
}
